src/XrdTpc/XrdTpcTPC.cc

```diff
@@ -160,43 +160,56 @@ int TPCHandler::closesocket_callback(void *clientp, curl_socket_t fd) {
 // We need to utilize the full URL (including the query string), not just the
 // resource name.  The query portion is hidden in the `xrd-http-query` header;
 // we take this out and combine it with the resource name.
+// We also append the value of the headers configured in tpc.header2cgi to the resource full URL
 //
 // One special key is `authz`; this is always stripped out and copied to the Authorization
 // header (which will later be used for XrdSecEntity).  The latter copy is only done if
 // the Authorization header is not already present.
 //
 // hasSetOpaque will be set to true if at least one opaque data has been set in the URL that is returned,
 // false otherwise
-static std::string prepareURL(XrdHttpExtReq &req, bool & hasSetOpaque) {
-    std::map<std::string, std::string>::const_iterator iter = req.headers.find("xrd-http-query");
-    if (iter == req.headers.end() || iter->second.empty()) {return req.resource;}
+std::string TPCHandler::prepareURL(XrdHttpExtReq &req, bool & hasSetOpaque) {
+    auto iter = req.headers.find("xrd-http-query");
+    bool found_first_header = false;
+    std::stringstream opaque;
 
-    auto has_authz_header = req.headers.find("authorization") != req.headers.end();
+    if (iter != req.headers.end() && !iter->second.empty()) {
+      auto has_authz_header = req.headers.find("authorization") != req.headers.end();
+      std::istringstream requestStream(iter->second);
+      std::string token;
 
-    std::istringstream requestStream(iter->second);
-    std::string token;
-    std::stringstream result;
-    bool found_first_header = false;
-    while (std::getline(requestStream, token, '&')) {
+      while (std::getline(requestStream, token, '&')) {
         if (token.empty()) {
-            continue;
+          continue;
         } else if (!strncmp(token.c_str(), "authz=", 6)) {
-            if (!has_authz_header) {
-                req.headers["authorization"] = token.substr(6);
-                has_authz_header = true;
-            }
-        } else if (!found_first_header) {
-            result << "?" << token;
-            found_first_header = true;
+          if (!has_authz_header) {
+            req.headers["authorization"] = token.substr(6);
+            has_authz_header = true;
+          }
         } else {
-            result << "&" << token;
+          opaque << (found_first_header ? "&" : "?") << token;
+          found_first_header = true;
+        }
+      }
+    }
+
+    // Append CGI coming from the tpc.header2cgi parameter
+    for(auto & hdr2cgi : hdr2cgimap) {
+        auto it = std::find_if(req.headers.begin(),req.headers.end(),[&hdr2cgi](const auto & elt){
+            // We exclude the Authorization header from the hdr to cgi conversion, as it will be appended later on to the opaque before the open
+            // happens
+            return elt.first != "authorization" && !strcasecmp(elt.first.c_str(),hdr2cgi.first.c_str());
```

Comment:
Maybe you need `strcasecmp(elt.first.c_str(), "authorization")` here, no?

Reply:
No no I cannot do this, Brian seems to use it ;) see https://github.com/xrootd/xrootd/issues/2222#issue-2203719451 the second line

Reply:
I'm not quite following the comment?

Indeed, we translate `Authorization` -> `authz` in many places.  However, what's the harm in allowing someone to set it again here?  Isn't the worst case that the token is appended twice?

Reply:
Yes, that's what would happen and I do not know if this is something we care about or not...

Reply:
Here's what we will get in the open opaque when I pass the `Authorization: Bearer test` header with my request and when the `tpc.header2cgi Authorization authz` is set: `"authz=Bearer test&nbstreams=10&oss.asize=5242880&authz=Bearer%20test"`

Reply:
Now with the 'authorization' exclusion:
```
nbstreams=10&oss.asize=5242880&authz=Bearer%20test"
```

Reply:
What I meant is that the map `tpc.header2cgi` should never contain `Authorization` as key, isn't that the intention in any case?

Reply:
Yes indeed, we can remove it from the map actually because if it is there, it will be automatically be appended to the opaque of the open no matter what...

Reply:
OK done. Anyway, this should not be merged until rebased on top of the case-insensitive header parsing commit